src/AutoRest.CSharp/Common/Input/Source/CodeGenAttributes.cs

```diff
@@ -0,0 +1,121 @@
+            CodeGenMemberSerializationHooksAttribute = GetSymbol(compilation, typeof(CodeGenMemberSerializationHooksAttribute));
+        }
+
+        public INamedTypeSymbol CodeGenSuppressAttribute { get; }
+
+        public INamedTypeSymbol CodeGenMemberAttribute { get; }
+
+        public INamedTypeSymbol CodeGenTypeAttribute { get; }
+
+        public INamedTypeSymbol CodeGenModelAttribute { get; }
+
+        public INamedTypeSymbol CodeGenClientAttribute { get; }
+
+        public INamedTypeSymbol CodeGenMemberSerializationAttribute { get; }
+
+        public INamedTypeSymbol CodeGenMemberSerializationHooksAttribute { get; }
+
+        private static INamedTypeSymbol GetSymbol(Compilation compilation, Type type) => compilation.GetTypeByMetadataName(type.FullName!) ?? throw new InvalidOperationException($"cannot load symbol of attribute {type}");
+
+        private static bool CheckAttribute(AttributeData attributeData, INamedTypeSymbol codeGenAttribute)
+            => SymbolEqualityComparer.Default.Equals(attributeData.AttributeClass, codeGenAttribute);
+
+        public bool TryGetCodeGenMemberAttributeValue(AttributeData attributeData, [MaybeNullWhen(false)] out string name)
+        {
+            name = null;
+            if (!CheckAttribute(attributeData, CodeGenMemberAttribute))
+                return false;
+
+            name = attributeData.ConstructorArguments.FirstOrDefault().Value as string;
+            return name != null;
+        }
+
+        public bool TryGetSerializationAttributeValue(AttributeData attributeData, [MaybeNullWhen(false)] out string[] propertyNames)
+        {
+            propertyNames = null;
+            if (!CheckAttribute(attributeData, CodeGenMemberSerializationAttribute))
+                return false;
+
+            if (attributeData.ConstructorArguments.Length > 0)
+            {
+                propertyNames = ToStringArray(attributeData.ConstructorArguments[0].Values);
+            }
+
+            return propertyNames != null;
+        }
+
+        public bool TryGetSerializationHooksAttributeValue(AttributeData attributeData, out (string? SerializationHook, string? DeserializationHook) hooks)
+        {
+            hooks = default;
+            if (!CheckAttribute(attributeData, CodeGenMemberSerializationHooksAttribute))
+                return false;
+
+            string? serializationHook = null;
+            string? deserializationHook = null;
+
+            var arguments = attributeData.ConstructorArguments;
+            serializationHook = arguments[0].Value as string;
+            deserializationHook = arguments[1].Value as string;
+
+            hooks = (serializationHook, deserializationHook);
+            return serializationHook != null || deserializationHook != null;
+        }
+
+        public bool TryGetCodeGenModelAttributeValue(AttributeData attributeData, out string[]? usage, out string[]? formats)
```

Comment:
Just curious about the best practise here, can we just return an empty array here instead of setting it as null?

Reply:
We should only return null when null has a meaning.
Here null has meanings for these two properties because they are patches to usage and formats, `null` means "do not patch"

Reply:
another reason is that this has been implemented this way therefore I do not really want to risk to break it since it is working properly.
All the things I have done is to move it here.

Reply:
I see.   Considering this is the existing behavour, let's keep it to avoid the risk of breaking.